smithy-model/src/main/java/software/amazon/smithy/model/traits/ExternalDocumentationTrait.java

```diff
@@ -17,37 +17,119 @@
 
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
 import software.amazon.smithy.model.SourceException;
 import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
 import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.Pair;
+import software.amazon.smithy.utils.ToSmithyBuilder;
 
 /**
- * Provides a link to external documentation of a service or operation.
+ * Provides named links to external documentation.
  */
-public final class ExternalDocumentationTrait extends StringTrait {
+public final class ExternalDocumentationTrait extends AbstractTrait
+        implements ToSmithyBuilder<ExternalDocumentationTrait> {
+
     public static final ShapeId ID = ShapeId.from("smithy.api#externalDocumentation");
 
-    public ExternalDocumentationTrait(String value, SourceLocation sourceLocation) {
-        super(ID, value, sourceLocation);
-        validateUrl(value, sourceLocation);
-    }
+    private final Map<String, String> urls;
 
-    public ExternalDocumentationTrait(String value) {
-        this(value, SourceLocation.NONE);
-    }
+    public ExternalDocumentationTrait(Builder builder) {
+        super(ID, builder.sourceLocation);
+        this.urls = Collections.unmodifiableMap(builder.urls);
 
-    public static final class Provider extends StringTrait.Provider<ExternalDocumentationTrait> {
-        public Provider() {
-            super(ID, ExternalDocumentationTrait::new);
+        if (urls.isEmpty()) {
```

Comment:
This validation could be removed and just handled by the model validation process if you add the `@length` trait to the map in the prelude model.